runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/HttpWatcherTask.java

```diff
@@ -0,0 +1,260 @@
+package io.aklivity.zilla.runtime.engine.internal.registry;
+
+import static java.net.http.HttpClient.Redirect.NORMAL;
+import static java.net.http.HttpClient.Version.HTTP_2;
+import static org.agrona.LangUtil.rethrowUnchecked;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+
+import io.aklivity.zilla.runtime.engine.config.NamespaceConfig;
+
+public class HttpWatcherTask extends WatcherTask
+{
+    private static final String INTIAL_ETAG = "INIT";
+    private static final URI CLOSE_REQUESTED = URI.create("http://localhost:12345");
+
+    private final Map<URI, String> etags;
+    private final Map<URI, byte[]> configHashes;
+    private final Map<URI, CompletableFuture<Void>> futures;
+    private final BlockingQueue<URI> configWatcherQueue;
+    private final ScheduledExecutorService executor;
+    //If server does not support long-polling use this interval
+    private final int pollIntervalSeconds;
+
+    public HttpWatcherTask(
+        BiFunction<URL, String, NamespaceConfig> changeListener,
```

Comment:
Feels like this `URL` type in `changeListener` is causing more exception handling than desired, like `MalformedURLException`.

Perhaps it would be better to use `URI` here instead to simplify the code and remove unnecessary try/catch blocks?

Reply:
If we change to URI here, I think we also should to change in a lot of places, e.g: in the guard to be consistent. If we only change it here, we're just delegating the exception handling to the readURL part, where we'd transform the URI to URL. But the readURL is dependent on the URL.openConnection method so I think we have to stick with URL there. If you think it's worth delegating the exception handling there, I'm up for it.

Reply:
Yeah, let's discuss the impact of changing it to `URI` instead of `URL`.